### ISSUE TITLE
Server should not know KV types on cache

### ISSUE THREAD

**Issue by thesreyas** [hazelcast/hazelcast#17522]
We want to use cache feature on hazelcast in client-server mode.
Server would start from hazelcast published docker.
KV type classes are available only on client side.
We donot want to deploy KV type classes on the server side 

Why does server needs to look for KV types on classpath? if it does not serve any functionality can you please skip that?

Below issue we see that hazelcast expects KV Types on classpath.
https://github.com/hazelcast/hazelcast/issues/16912

**Comment by jerrinot** [comment]
hi @thesreyas,

what Hazelcast version do you use? The issue you mentioned is fixed in Hazelcast [3.12.8](https://github.com/hazelcast/hazelcast/issues/16912) and [4.0.2](https://github.com/hazelcast/hazelcast/pull/16918)

Can you describe the failing scenario a bit?

**Comment by vbekiaris** [comment]
@thesreyas there are certain features (like executing `EntryProcessor`s or storing items in `OBJECT` format) which require objects in the cache to be deserialized in order to be processed on the member side. However with the default `BINARY` object format it is absolutely possible to use a `Cache` from client side without requiring presence of domain object classes on classpath.

The issue you linked to is a bug and is fixed in 3.12.8. Apart from that, can you be more specific about issues you found when trying to use cache from the client side?

**Comment by shashanka981** [comment]
> @thesreyas there are certain features (like executing `EntryProcessor`s or storing items in `OBJECT` format) which require objects in the cache to be deserialized in order to be processed on the member side. However with the default `BINARY` object format it is absolutely possible to use a `Cache` from client side without requiring presence of domain object classes on classpath.

Is this true for Javax.cache as well ? 
[CacheCreationIssueHz.zip](https://github.com/hazelcast/hazelcast/files/5239383/CacheCreationIssueHz.zip)


Using Hazelcast 3.12.8 , i am creating a javax.cache from hazelcast client without deploying type definitions on server member and i see cache creation fails with below exception. Attached the server and client classes that i used to reproduce the issue . As soon as i add the Type definition of server, cache creation succeeds


INFO: testInstance [dev] [3.12.8] Diagnostics disabled. To enable add -Dhazelcast.diagnostics.enabled=true to the JVM arguments.
Exception in thread "main" com.hazelcast.nio.serialization.HazelcastSerializationException: java.lang.ClassNotFoundException: com.pega.ClassA
	at com.hazelcast.internal.serialization.impl.JavaDefaultSerializers$ClassSerializer.read(JavaDefaultSerializers.java:291)
	at com.hazelcast.internal.serialization.impl.JavaDefaultSerializers$ClassSerializer.read(JavaDefaultSerializers.java:279)
	at com.hazelcast.internal.serialization.impl.StreamSerializerAdapter.read(StreamSerializerAdapter.java:48)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.readObject(AbstractSerializationService.java:269)
	at com.hazelcast.internal.serialization.impl.ByteArrayObjectDataInput.readObject(ByteArrayObjectDataInput.java:574)
	at com.hazelcast.config.CacheConfig.readKeyValueTypes(CacheConfig.java:658)
	at com.hazelcast.config.CacheConfig.readData(CacheConfig.java:564)
	at com.hazelcast.internal.serialization.impl.DataSerializableSerializer.readInternal(DataSerializableSerializer.java:160)
	at com.hazelcast.internal.serialization.impl.DataSerializableSerializer.read(DataSerializableSerializer.java:106)
	at com.hazelcast.internal.serialization.impl.DataSerializableSerializer.read(DataSerializableSerializer.java:51)
	at com.hazelcast.internal.serialization.impl.StreamSerializerAdapter.read(StreamSerializerAdapter.java:48)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toObject(AbstractSerializationService.java:187)
	at com.hazelcast.spi.impl.NodeEngineImpl.toObject(NodeEngineImpl.java:324)
	at com.hazelcast.client.impl.protocol.task.cache.CacheCreateConfigMessageTask.extractCacheConfigFromMessage(CacheCreateConfigMessageTask.java:87)
	at com.hazelcast.client.impl.protocol.task.cache.CacheCreateConfigMessageTask.processMessage(CacheCreateConfigMessageTask.java:56)
	at com.hazelcast.client.impl.protocol.task.AbstractMessageTask.initializeAndProcessMessage(AbstractMessageTask.java:129)
	at com.hazelcast.client.impl.protocol.task.AbstractMessageTask.run(AbstractMessageTask.java:109)
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:163)
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.process(OperationThread.java:159)
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.process(OperationThread.java:127)
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.run(OperationThread.java:110)
	at ------ submitted from ------.(Unknown Source)
	at com.hazelcast.client.spi.impl.ClientInvocationFuture.resolveAndThrowIfException(ClientInvocationFuture.java:96)
	at com.hazelcast.client.spi.impl.ClientInvocationFuture.resolveAndThrowIfException(ClientInvocationFuture.java:33)
	at com.hazelcast.spi.impl.AbstractInvocationFuture.get(AbstractInvocationFuture.java:163)
	at com.hazelcast.client.cache.impl.ClientCacheHelper.createCacheConfig(ClientCacheHelper.java:126)
	at com.hazelcast.client.cache.impl.HazelcastClientCacheManager.createCacheConfig(HazelcastClientCacheManager.java:144)
	at com.hazelcast.cache.impl.AbstractHazelcastCacheManager.createCacheInternal(AbstractHazelcastCacheManager.java:119)
	at com.hazelcast.cache.impl.AbstractHazelcastCacheManager.createCache(AbstractHazelcastCacheManager.java:147)
	at com.hazelcast.cache.impl.AbstractHazelcastCacheManager.createCache(AbstractHazelcastCacheManager.java:65)
	at com.pega.StartHazelcastClient.createCache(StartHazelcastClient.java:60)
	at com.pega.StartHazelcastClient.startClient(StartHazelcastClient.java:53)
	at com.pega.StartHazelcastClient.main(StartHazelcastClient.java:39)
Caused by: java.lang.ClassNotFoundException: com.pega.ClassA
	at jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:581)
	at jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:178)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:522)
	at com.hazelcast.nio.ClassLoaderUtil.tryLoadClass(ClassLoaderUtil.java:288)
	at com.hazelcast.nio.ClassLoaderUtil.loadClass(ClassLoaderUtil.java:252)

**Comment by vbekiaris** [comment]
@shashanka981 I tested your code and works fine for me, maybe you can describe more about your environment so we can reproduce your issue? There are a few minor things to fix in your sample code (that actually seem unrelated to the issue at hand):

1. Moved your `config.setClassLoader` statements before starting the `HazelcastInstance. As indicated in the [reference manual](https://docs.hazelcast.org/docs/4.0.2/manual/html-single/index.html#configuring-programmatically), programmatic configuration should not be modified after the instance is started; otherwise this might lead to unexpected behaviour.
2. Removed the client network config for address as it is not applicable for my environment.
3. Assumed a `public class ClassA implements Serializable { public String getA() {...} }` as a placeholder for your actual class.
4. Changed usage of `HazelcastClientCachingProvider.createCachingProvider(client)` which is directly invoking internal API. If you want to reference an existing instance, you can do so by instance name by supplying a property to standard JCache API methods:
```
// the client-side provider is default when HazelcastClient is in classpath
// alternatively you can set Hazelcast property hazelcast.jcache.provider.type
// to a value supported by your Hazelcast version ("server", "client")
CachingProvider provider = Caching.getCachingProvider(); 
// see also https://docs.hazelcast.org/docs/4.0.2/manual/html-single/index.html#binding-to-a-named-instance
CacheManager manager = provider.getCacheManager(provider.getDefaultURI(),
                    provider.getDefaultClassLoader(),
                    HazelcastCachingProvider.propertiesByInstanceName("testInstance"));
```

Notice that all reference manual links above point to latest published version (4.0.2), please check ones that match your Hazelcast version.

**Comment by mmedenjak** [comment]
@shashanka981 have you managed to take a look at the suggestions above and comment on your usage?

**Comment by shashanka981** [comment]
@vbekiaris 
I hope the ClassA definition you have added during you testing is not available to Server member. Am I right ? 

Regarding details of environment, i am on Ubuntu and using Open JDK "1.8.0_222". Not sure what other information you are looking for ? I do not have any other configuration in place other than the files what I already shared with you. BTW I tested my changes again with the minor issues you pointed and I still see cache creation fails when server member doesnt have access to type definition of ClassA

**Comment by mmedenjak** [comment]
@shashanka981 can you package your project and send it to us so we take a quick look? Maybe we can resolve the issue quicker.

**Comment by shashanka981** [comment]
Please find attached projects - one for client and one for server
[HazelcastTestClient.zip](https://github.com/hazelcast/hazelcast/files/5256091/HazelcastTestClient.zip)
[HazelcastTestServer.zip](https://github.com/hazelcast/hazelcast/files/5256096/HazelcastTestServer.zip)

**Comment by vbekiaris** [comment]
@shashanka981 thanks for the projects. I verified your project fails when the value class is missing from the server-side. This is due to a combination of:
  - your `CacheConfig` explicitly setting cache KV types.
  - using Hazelcast 3.x series. The serialized form of `CacheConfig` when sent from client to server in 3.x includes the serialized `Class`es of the KV types, when those are set in any way (using any of `setTypes`, `setKeyClassName`, `setValueClassName`, `setKeyType`, `setValueType` on a `CacheConfig`). The serialized form of `CacheConfig` cannot be changed in the 3.x series due to the client compatibility guarantee: starting with Hazelcast 3.6 any client can be used with a cluster of members in any 3.6 or later version.

Probably I only tested using member & client of the latest released version (4.0.2). Since 4.x, client uses a different serialized form for `CacheConfig` that only sends the names of the KV types to the server, thus there is no requirement for the KV classes to be on the server's classpath. Runtime type validation happens on the client-side.

Your options are:
- if you want to use 3.12.8, don't set KV types on your `CacheConfig`. You will still have the compile-time checks when using the parameterized `Cache<String, ClassA>`, without the runtime type checks.
- upgrade to 4.x

As a sidenote, JCache API latest version is 1.1.1, make sure you use this artifact with Hazelcast 3.12.8 or 4.x. See the documentation related to JCache versions & Hazelcast versions [here](https://docs.hazelcast.org/docs/4.0.3/manual/html-single/index.html#supported-jcache-versions).

**Comment by shashanka981** [comment]
@vbekiaris Thanks for the info. 
Your proposed Option 1) Is it applicable for 3.11.5  Apart from avoiding run time type checks , are there any other side effects that comes with Not setting the KV Types on CacheConfig ?

**Comment by vbekiaris** [comment]
@shashanka981 if you want to stay on the 3.x series, I would suggest you use the latest version (which currently is 3.12.9). Version 3.11.5 does not include the fix for #16912 (which is mentioned in the first comment in this issue), it is fixed since 3.12.8.

Not setting the KV types on the `CacheConfig` skips the runtime type checks and has no other significant side effects. There is a minor related behavioural change in JCache API. I don't expect it affects you, but mention it here for completeness (it is also explained in the [previous JCache/Hazelcast ref manual link](https://docs.hazelcast.org/docs/4.0.3/manual/html-single/index.html#supported-jcache-versions) I posted):
- as of JCache 1.0.0, if you create a cache with specific KV types, then `CacheManager.getCache(String name)` will not return that `Cache`. The only way to obtain a reference to an existing `Cache` with configured KV types is to use `CacheManager#getCache(String cacheName, Class<K> keyType, Class<V> valueType)`.
- since JCache 1.1.0 (Hazelcast 3.9.3+), the JCache API specification relaxed `CacheManager.getCache(String name)` to not perform KV types checking. So it will return a `Cache` that matches by name, even when explicit KV types are configured for that cache.

**Comment by vbekiaris** [comment]
Closing this issue now, @shashanka981 if you have further questions feel free to reach out on [google group](https://groups.google.com/forum/#!forum/hazelcast) or [community slack](https://slack.hazelcast.com/)
